File: ios/TransitKit/Sources/Components/LineBadge.swift
import SwiftUI

// MARK: - LineBadge
//
// Single source of truth for line badges across the app. Every place that
// renders a GTFS line identifier — departure rows, line lists, stop
// coincidences, headers, filter chips, vehicle cards, trip details —
// goes through this component.
//
// Map annotations (VehicleAnnotationView, StopAnnotationView) are NOT in
// scope: those use custom shapes (dots, pins with halos) that aren't
// badges. Keep them as-is.
//
// Design is mono-operator. The "operator logo circle" variant from
// movete is intentionally omitted — every TransitKit install serves a
// single agency, so it would always be redundant. The API keeps
// `showTransitIcon` so multi-modal operators (bus + rail + ferry) can
// opt in to an icon chip.

/// Size preset — drives font, padding, min width, corner radius and icon size.
enum LineBadgeSize {
    case small   // coincidences, filter chips, dense lists
    case medium  // cards, sheet headers, secondary rows
    case large   // departure rows, line lists, trip headers

    var fontSize: CGFloat {
        switch self {
        case .small:  11
        case .medium: 13
        case .large:  15
        }
    }

    var iconSize: CGFloat {
        switch self {
        case .small:  12
        case .medium: 14
        case .large:  16
        }
    }

    var hPadding: CGFloat {
        switch self {
        case .small:  6
        case .medium: 8
        case .large:  10
        }
    }

    var vPadding: CGFloat {
        switch self {
        case .small:  3
        case .medium: 4
        case .large:  5
        }
    }

    var minWidth: CGFloat {
        switch self {
        case .small:  24
        case .medium: 32
        case .large:  40
        }
    }

    var cornerRadius: CGFloat {
        switch self {
        case .small:  4
        case .medium: 6
        case .large:  6
        }
    }

    var spacing: CGFloat {
        switch self {
        case .small:  3
        case .medium: 5
        case .large:  6
        }
    }
}

/// Pill-shaped badge showing a transit line name with GTFS route color.
///
/// ```swift
/// // Departure / route convenience — pulls color + text + transitType
/// LineBadge(departure: dep, size: .large)
/// LineBadge(route: apiRoute, size: .medium)
///
/// // Low-level — explicit color strings (no Route/Departure handy)
/// LineBadge(name: "R", color: "#AD2B3C", textColor: nil, transitType: .bus)
///
/// // Opt-in transit icon (bus/tram/rail/ferry)
/// LineBadge(route: apiRoute, size: .large, showTransitIcon: true)
/// ```
struct LineBadge: View {
    let name: String
    /// GTFS background color as hex (e.g. "#AD2B3C").
    let color: String
    /// GTFS text color as hex, or nil to auto-derive by WCAG luminance.
    let textColor: String?
    /// Optional transit type for the icon chip. Ignored when `showTransitIcon=false`.
    let transitType: TransitType?
    var size: LineBadgeSize = .medium
    var showTransitIcon: Bool = false

    // MARK: - Inits

    /// Convenience from a `Departure`. Defaults to `.large` (primary context).
    init(departure: Departure, size: LineBadgeSize = .large, showTransitIcon: Bool = false) {
        self.name = departure.lineName
        self.color = departure.color
        self.textColor = departure.textColor
        self.transitType = departure.transitType
        self.size = size
        self.showTransitIcon = showTransitIcon
    }

    /// Convenience from an `APIRoute`. Defaults to `.large`.
    init(route: APIRoute, size: LineBadgeSize = .large, showTransitIcon: Bool = false) {
        self.name = route.name
        self.color = route.color ?? "#000000"
        self.textColor = route.textColor
        self.transitType = route.resolvedTransitType
        self.size = size
        self.showTransitIcon = showTransitIcon
    }

    /// Low-level init for contexts without a Route/Departure model.
    /// Pass `textColor: nil` to auto-derive by WCAG luminance.
    init(
        name: String,
        color: String,
        textColor: String? = nil,
        transitType: TransitType? = nil,
        size: LineBadgeSize = .medium,
        showTransitIcon: Bool = false
    ) {
        self.name = name
        self.color = color
        self.textColor = textColor
        self.transitType = transitType
        self.size = size
        self.showTransitIcon = showTransitIcon
    }

    // MARK: - Body

    var body: some View {
        HStack(spacing: size.spacing) {
            if showTransitIcon, let type = transitType {
                type.icon.sized(size.iconSize)
                    .foregroundStyle(fgColor)
            }
            Text(name)
                .font(.system(size: size.fontSize, weight: .bold))
                .foregroundStyle(fgColor)
                .lineLimit(1)
                .fixedSize(horizontal: true, vertical: false)
        }
        .padding(.horizontal, size.hPadding)
        .padding(.vertical, size.vPadding)
        .frame(minWidth: size.minWidth)
        .background(bgColor, in: RoundedRectangle(cornerRadius: size.cornerRadius, style: .continuous))
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(String(format: NSLocalizedString("line_badge_a11y", comment: ""), name))
    }

    // MARK: - Color resolution

    private var bgColor: Color { Color(hex: color) }
    private var fgColor: Color { Color(hex: Self.resolvedTextColor(textColor, background: color)) }

    /// Text color rules:
    /// - `nil`, empty, or WCAG-sentinel values ("000000"/"FFFFFF") → always
    ///   compute by luminance (GTFS feeds often fill these as placeholders).
    /// - Any other provided value is kept only if it passes WCAG 4.5:1 against
    ///   the background. Otherwise fall back to the computed contrast color.
    private static func resolvedTextColor(_ textColor: String?, background: String) -> String {
        let sentinels: Set<String> = ["", "#", "000000", "FFFFFF", "#000000", "#FFFFFF"]
        guard let textColor, !sentinels.contains(textColor) else {
            return contrastingTextColor(for: background)
        }
        let candidate = textColor.hasPrefix("#") ? textColor : "#\(textColor)"
        let bg = background.hasPrefix("#") ? background : "#\(background)"
        func lum(_ hex: String) -> Double {
            func lin(_ c: Double) -> Double { c <= 0.04045 ? c/12.92 : pow((c+0.055)/1.055, 2.4) }
            let h = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
            var v: UInt64 = 0; Scanner(string: h).scanHexInt64(&v)
            guard h.count == 6 else { return 0 }
            return 0.2126*lin(Double((v>>16)&0xFF)/255)
                 + 0.7152*lin(Double((v>>8)&0xFF)/255)
                 + 0.0722*lin(Double(v&0xFF)/255)
        }
        let Lbg = lum(bg); let Lfg = lum(candidate)
        let ratio = Lbg > Lfg ? (Lbg+0.05)/(Lfg+0.05) : (Lfg+0.05)/(Lbg+0.05)
        return ratio >= 4.5 ? candidate : contrastingTextColor(for: background)
    }
}

File: android/app/src/main/java/com/transitkit/app/ui/components/LineBadge.kt
package com.transitkit.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.transitkit.app.config.LucideIcons
import com.transitkit.app.data.model.ScheduleRoute
import com.transitkit.app.data.model.ResolvedDeparture

// -----------------------------------------------------------------------------
// LineBadge — the only component that renders a GTFS line pill in the app.
// -----------------------------------------------------------------------------
// Same design as the iOS sibling. Scope: departure rows, line lists, stop
// coincidences, filter chips, sheet headers, vehicle/trip cards.
//
// Map annotations (VehicleAnnotationView, StopAnnotationView) are NOT in
// scope: they use custom shapes (dot + halo, pin + tail) and stay as-is.
//
// Single-operator-per-app product → no "operator logo" chip variant.
// `showTransitIcon` toggles an optional left-side modal icon (bus / tram /
// rail / ferry) for multi-modal agencies.
// -----------------------------------------------------------------------------

enum class LineBadgeSize { Small, Medium, Large }

private data class LineBadgeDims(
    val fontSize: TextUnit,
    val iconSize: Dp,
    val padH: Dp,
    val padV: Dp,
    val minWidth: Dp,
    val radius: Dp,
    val spacing: Dp,
)

private fun LineBadgeSize.dims(): LineBadgeDims = when (this) {
    LineBadgeSize.Small -> LineBadgeDims(11.sp, 12.dp, 6.dp, 3.dp, 24.dp, 4.dp, 3.dp)
    LineBadgeSize.Medium -> LineBadgeDims(13.sp, 14.dp, 8.dp, 4.dp, 32.dp, 6.dp, 5.dp)
    LineBadgeSize.Large -> LineBadgeDims(15.sp, 16.dp, 10.dp, 5.dp, 40.dp, 6.dp, 6.dp)
}

/**
 * Convenience overload from a [ScheduleRoute]. Defaults to [LineBadgeSize.Large]
 * (primary context: lines list, line detail header, departure rows).
 */
@Composable
fun LineBadge(
    route: ScheduleRoute,
    size: LineBadgeSize = LineBadgeSize.Large,
    showTransitIcon: Boolean = false,
    modifier: Modifier = Modifier,
) {
    LineBadge(
        name = route.name,
        colorHex = route.color,
        textColorHex = route.textColor,
        transitType = route.transitType,
        size = size,
        showTransitIcon = showTransitIcon,
        modifier = modifier,
    )
}

/**
 * Convenience overload from a [ResolvedDeparture]. Defaults to [LineBadgeSize.Large].
 */
@Composable
fun LineBadge(
    departure: ResolvedDeparture,
    size: LineBadgeSize = LineBadgeSize.Large,
    showTransitIcon: Boolean = false,
    modifier: Modifier = Modifier,
) {
    LineBadge(
        name = departure.routeName,
        colorHex = departure.routeColor,
        textColorHex = departure.routeTextColor,
        transitType = departure.transitType,
        size = size,
        showTransitIcon = showTransitIcon,
        modifier = modifier,
    )
}

/**
 * Low-level badge — expects hex color strings. Pass [textColorHex] = null to
 * auto-derive by WCAG luminance.
 */
@Composable
fun LineBadge(
    name: String,
    colorHex: String?,
    textColorHex: String? = null,
    transitType: Int? = null,
    size: LineBadgeSize = LineBadgeSize.Medium,
    showTransitIcon: Boolean = false,
    modifier: Modifier = Modifier,
) {
    val d = size.dims()
    val bg = parseHexColor(colorHex, fallback = Color(0xFF3B82F6))
    val fg = resolveTextColor(textColorHex, background = bg)
    val cd = "Linea $name"

    Row(
        modifier = modifier
            .defaultMinSize(minWidth = d.minWidth)
            .background(bg, RoundedCornerShape(d.radius))
            .padding(horizontal = d.padH, vertical = d.padV)
            .semantics { contentDescription = cd },
        horizontalArrangement = Arrangement.spacedBy(d.spacing, Alignment.CenterHorizontally),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (showTransitIcon) {
            transitIcon(transitType)?.let { iconRes ->
                Icon(
                    painter = painterResource(iconRes),
                    contentDescription = null,
                    tint = fg,
                    modifier = Modifier.size(d.iconSize),
                )
            }
        }
        androidx.compose.material3.Text(
            text = name,
            color = fg,
            fontSize = d.fontSize,
            fontWeight = FontWeight.Bold,
            maxLines = 1,
        )
    }
}

// -----------------------------------------------------------------------------
// Color helpers
// -----------------------------------------------------------------------------

/**
 * Parse a GTFS hex color ("#AABBCC" or "AABBCC"). Falls back to [fallback]
 * on any parse failure — feeds sometimes publish blank strings.
 */
fun parseHexColor(hex: String?, fallback: Color = Color(0xFF3B82F6)): Color {
    val s = hex?.trim()?.removePrefix("#") ?: return fallback
    if (s.length != 6 && s.length != 8) return fallback
    return runCatching {
        val argb = if (s.length == 6) "FF$s" else s
        Color(argb.toLong(16).toInt())
    }.getOrDefault(fallback)
}

/**
 * WCAG 2.1 relative luminance of a color.
 */
private fun Color.luminance(): Double {
    fun linearize(c: Float): Double {
        val v = c.toDouble()
        return if (v <= 0.03928) v / 12.92 else Math.pow((v + 0.055) / 1.055, 2.4)
    }
    return 0.2126 * linearize(red) + 0.7152 * linearize(green) + 0.0722 * linearize(blue)
}

/**
 * Returns a text color with WCAG-compliant contrast:
 * - Null / sentinel ("000000"/"FFFFFF") → auto-derive from background luminance.
 * - Provided value kept only if it passes 4.5:1 ratio; otherwise overridden.
 */
private fun resolveTextColor(textColorHex: String?, background: Color): Color {
    val sentinels = setOf("", "000000", "FFFFFF", "000", "FFF")
    val cleaned = textColorHex?.trim()?.removePrefix("#")?.uppercase()
    if (cleaned.isNullOrEmpty() || cleaned in sentinels) {
        return if (background.luminance() > 0.5) Color.Black else Color.White
    }
    val candidate = parseHexColor(cleaned, fallback = Color.White)
    val lBg = background.luminance()
    val lFg = candidate.luminance()
    val ratio = if (lBg > lFg) (lBg + 0.05) / (lFg + 0.05) else (lFg + 0.05) / (lBg + 0.05)
    return if (ratio >= 4.5) candidate
        else if (lBg > 0.5) Color.Black else Color.White
}

/**
 * Optional transit modal icon. Returns null when [transitType] is missing or
 * unmapped — the badge will just render as a color pill without icon.
 * GTFS route_type mapping per spec.
 */
private fun transitIcon(transitType: Int?): Int? = when (transitType) {
    0 -> LucideIcons.Tram          // 0 = Tram, Streetcar, Light rail
    1, 2, 12 -> LucideIcons.Train  // 1 = Subway/Metro, 2 = Rail, 12 = Monorail
    3, 11 -> LucideIcons.BusFront       // 3 = Bus, 11 = Trolleybus
    4 -> LucideIcons.Ship          // 4 = Ferry
    else -> null
}
